### PR TITLE
solver: do not punish explicitly requested compiler mismatches

### PR DESCRIPTION
**Pull request by becker33**
Fixes a bug reported by @haampie on slack. Includes regression test.

### CODE REVIEW

**Review by tgamblin — DISMISSED** [review]

**Review by alalazo — DISMISSED** [review]
This changes the semantics we had before. Something like:
```console
$ hdf5 ^mpich%clang 
```
will use the default compiler for anything above `mpich`. Not sure what would happen in more complex cases with more compilers set. See #22718 for a regression this introduces.

**Comment by becker33** [comment]
Yep, I need to modify this to punish mismatches that are required by set compilers, just at a lower priority than unsolicited mismatches.

**Comment by becker33** [comment]
@alalazo this should be fixed for the issue you highlighted

**Review by tgamblin — APPROVED** [review]

File: lib/spack/spack/solver/concretize.lp [hunk 1]
```diff
@@ -1291,6 +1299,15 @@ opt_criterion(40, "compiler mismatches").
       build_priority(Package, Priority)
 }.
 
+opt_criterion(39, "compiler mismatches that are not from CLI").
```
Comment (alalazo) [inline lib/spack/spack/solver/concretize.lp:1302]: I think these are from CLI?

**Comment by alalazo** [comment]
Compared 52cc798948e41eaa30dd7645bfa2d1ecda5f0fd3 (orange) with 258edf7dac07710ea6c50316dc53e5154d9cb3b7 (blue) on a `--fresh` concretization of the packages in RADIUSS:

[radiuss_develop.csv](https://github.com/spack/spack/files/11081243/radiuss_develop.csv)
[radiuss_30074.csv](https://github.com/spack/spack/files/11081244/radiuss_30074.csv)
[radiuss.txt](https://github.com/spack/spack/files/11081245/radiuss.txt)

The plot is:

totals

Apparently this increased the solve time for a few packages (the most noticeable is `visit`). The setup time increase for `glvis` is very likely noise.

**Pull request merged**